plugin/pulpcore/plugin/download/asyncio/base.py

```diff
@@ -0,0 +1,153 @@
+    instantiator to define the file to receive data allows the streamer to receive the data instead
+    of having it written to disk.
+    """
+
+    def __init__(self, custom_file_object=None, expected_digests=None, expected_size=None):
+        """
+        Create a BaseDownloader object. This is expected to be called by all subclasses.
+
+        Args:
+            custom_file_object (file object): An open, writable file object that downloaded data
+                can be written to by
+                :meth:`~pulpcore.plugin.download.asyncio.BaseDownloader.handle_data`.
+            expected_digests (dict): Keyed on the algorithm name provided by hashlib and stores the
+                value of the expected digest. e.g. {'md5': '912ec803b2ce49e4a541068d495ab570'}
+            expected_size (int): The number of bytes the download is expected to have.
+        """
+        if custom_file_object:
+            self._writer = custom_file_object
+        else:
+            fd, self.path = tempfile.mkstemp(dir=os.getcwd())
+            self._writer = os.fdopen(fd, mode='wb')
+        self.expected_digests = expected_digests
+        self.expected_size = expected_size
+        self._digests = {n: hashlib.new(n) for n in Artifact.DIGEST_FIELDS}
+        self._size = 0
+
+    def handle_data(self, data):
+        """
+        Write data to the file object and compute its digests.
+
+        All subclassed downloaders are expected to pass all data downloaded to this method. Similar
+        to the hashlib docstring, repeated calls are equivalent to a single call with
+        the concatenation of all the arguments: m.handle_data(a); m.handle_data(b) is equivalent to
+        m.handle_data(a+b).
+
+        Args:
+            data (bytes): The data to be handled by the downloader.
+        """
+        self._writer.write(data)
+        self._record_size_and_digests_for_data(data)
+
+    def _record_size_and_digests_for_data(self, data):
+        """
+        Record the size and digest for an available chunk of data.
+
+        Args:
+            data (bytes): The data to have its size and digest values recorded.
+        """
+        for algorithm in self._digests.values():
+            algorithm.update(data)
+        self._size = self._size + len(data)
+
+    @property
+    def artifact_attributes(self):
+        """
+        A property that returns a dictionary with size and digest information. The keys of this
+        dictionary correspond with :class:`pulpcore.plugin.models.Artifact` fields.
+        """
+        attributes = {'size': self._size}
+        for algorithm in Artifact.DIGEST_FIELDS:
+            attributes[algorithm] = self._digests[algorithm].hexdigest()
+        return attributes
+
+    def validate_digests(self):
+        """
+        Validate all digests validate if ``expected_digests`` is set
+
+        Raises:
+            :class:`~pulpcore.plugin.download.asyncio.DigestValidationError`: When any of the
```

Comment:
What value does the `:class:` syntax add?  Seems autodoc already links the class by just specifying the type.

Reply:
I think this is how Sphinx needs it. When I compile the docs without the `:class:` they don't link or look correctly (like Sphinx recognizes them). This is the syntax also they recommend which has `:class:`, `:attr:`, `:meth:`, etc. I'm leaving this as-is.